xmtp/src/builder.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{
+    client::{Client, Network},
+    persistence::{InMemoryPersistence, Persistence},
+};
+
+#[derive(Default)]
+pub struct ClientBuilder<P>
+where
+    P: Persistence,
+{
+    network: Network,
+    persistence: Option<P>,
+}
+
+impl<P> ClientBuilder<P>
+where
+    P: Persistence,
+{
+    pub fn new() -> Self {
+        Self {
+            network: Network::Dev,
+            persistence: None,
+        }
+    }
+
+    pub fn network(mut self, network: Network) -> Self {
+        self.network = network;
+        self
+    }
+
+    pub fn persistence(mut self, persistence: P) -> Self {
+        self.persistence = Some(persistence);
+        self
+    }
+
+    pub fn build(self) -> Client<P> {
+        Client {
+            network: self.network,
+            persistence: self.persistence.expect("A persistence engine must be set"),
```

Comment:
Theres an open question of whether the persistence engine should initialized in New, which allows for Compile time check, or if it should be lazily initialized on build which allows more flexibility. 

went with flexibility for now

Reply:
Adding Default trait to Persistence would allow the builder to instantiate on build and avoid a panic, but seems heavy weight at this time.

Reply:
Default for persistence would just be choosing an engine like `InMemoryPersistence` right? It makes sense to just panic to me. The consumer should be forced to acknowledge the persistence method they want to use.

Reply:
The Default trait would provide a function called default which accepts no params. Adding this to the persistence trait would force every struct who implements Persistence to have a default initialization. 

For InMem likely makes sense, however for more complicated Persistence implementation a default might not make sense (a hypothetical HttpPersistence would minimally need a url to be useful). 

Ultimately it would be great to throw a compile time error if a value is missing, but I Figure it doesn't need to be decided right meow...